contract/src/main/kotlin/io/provenance/scope/loan/utility/LoanContractValidations.kt

```diff
@@ -55,41 +57,79 @@ internal val eNoteControllerValidation: ContractEnforcementContext.(ENoteControl
-        document.checksum.isValid()        orError "ENote missing checksum",
+        document.id.isValid()              orError "ENote must have valid ID",
+        document.uri.isNotBlank()          orError "ENote is missing URI",
+        document.contentType.isNotBlank()  orError "ENote is missing content type",
+        document.documentType.isNotBlank() orError "ENote is missing document type",
+        document.checksum.isValid()        orError "ENote is missing checksum",
     )
 }
 
+internal val eNoteInputValidation: (ENote) -> Unit = { eNote ->
+    validateRequirements(ContractRequirementType.VALID_INPUT) {
+        eNoteValidation(eNote)
+    }
+}
+
 internal val eNoteValidation: ContractEnforcementContext.(ENote) -> Unit = { eNote ->
     // TODO: Decide which fields should only be required if DART is listed as mortgagee of record/active custodian
     eNoteControllerValidation(eNote.controller)
     eNoteDocumentValidation(eNote.eNote)
     requireThat(
-        eNote.signedDate.isValid()   orError "ENote missing signed date",
-        eNote.vaultName.isNotBlank() orError "ENote missing vault name",
+        eNote.signedDate.isValid()   orError "ENote is missing signed date",
+        eNote.vaultName.isNotBlank() orError "ENote is missing vault name",
     )
 }
 
+internal val loanDocumentInputValidation: (LoanDocuments) -> Unit = { loanDocuments ->
+    validateRequirements(ContractRequirementType.VALID_INPUT) {
+        requireThat(
+            loanDocuments.documentList.isNotEmpty() orError "Must supply at least one document" // TODO: Verify desired; not thrown for optional input
+        )
+        val incomingDocChecksums = mutableMapOf<String, Boolean>()
+        loanDocuments.documentList.forEach { document ->
+            documentValidation(document)
+            document.checksum.takeIf { it.isValid() }?.checksum?.let { checksum ->
+                if (incomingDocChecksums[checksum] == true) {
+                    raiseError("Loan document with checksum $checksum already provided in input")
+                }
+                incomingDocChecksums[checksum] = true
+            }
+        }
+    }
+}
+
 internal val loanStateValidation: ContractEnforcementContext.(LoanStateMetadata) -> Unit = { loanState ->
     val idSnippet = if (loanState.id.isSet()) {
         " with ID ${loanState.id.value}"
     } else {
         ""
     }
     requireThat(
-        loanState.id.isValid()              orError "Loan state missing valid ID",
-        loanState.effectiveTime.isValid()   orError "Loan state$idSnippet missing effective time",
-        loanState.uri.isNotBlank()          orError "Loan state$idSnippet missing URI",
-        loanState.checksum.isValid()        orError "Loan state$idSnippet missing checksum"
+        loanState.id.isValid()              orError "Loan state must have valid ID",
+        // TODO: Validity check for effective time originally included "< Instant.now()", confirm if that's really desired or not
```

Comment:
I'd say keep it.. we don't do much data bounding, but this is an obvious one to prevent servicers from accidentally recording loan tapes that are future dated. We could potentially do a lower bound of the closing/signed date, too.